### ISSUE TITLE
v0.4 Issue - Launch button does not seem to work (Linux, Steam)

### ISSUE THREAD

**Issue by Slayer5934** [Nexus-Mods/NexusMods.App#1151]
# Bug Report

## Summary
When trying to launch Stardew the Launch button does not seem to work. It says Running... then eventually turns back to Launch.

## Steps to reproduce
...

## What is the expected behaviour?
It launches.

## Other information
It is worth noting that I installed SMAPI (and other mods) outside of the Nexus Mods App before getting it.
I installed one mod, applied, then launched through steam directly and everything worked fine.

**Comment by erri120** [comment]
Are you on Windows or on Linux? Also, can you attach some logs?

Windows: `%LOCALAPPDATA%\NexusMods.App\Logs`
Linux: `~/.local/state/NexusMods.App/Logs`

**Comment by Slayer5934** [comment]
[nexusmods.app.current.log](https://github.com/Nexus-Mods/NexusMods.App/files/14827976/nexusmods.app.current.log)

I do not see anything in the log that would suggest a problem... I could run a dev build or run in debug mode or something if you needed it though

**Comment by erri120** [comment]
> [nexusmods.app.current.log](https://github.com/Nexus-Mods/NexusMods.App/files/14827976/nexusmods.app.current.log)
> 
> I do not see anything in the log that would suggest a problem... I could run a dev build or run in debug mode or something if you needed it though

You can try running this command manually:

```bash
xdg-open steam://rungameid/413150
```

On Linux, we currently only support running the game through Steam. The only thing the launch button does is run the command above.

https://github.com/Nexus-Mods/NexusMods.App/blob/5be72d23ea133fac8d5b4f114656b63c48aeb3f8/src/Abstractions/NexusMods.Abstractions.Games/RunGameTool.cs#L116-L142

**Comment by Slayer5934** [comment]
`xdg-open steam://rungameid/413150` works as expected
Running through the mod manager doesn't seem to work, not sure why, it says running tool so perhaps its because I installed Smapi outside of the manager? Just trying to guess
![image](https://github.com/Nexus-Mods/NexusMods.App/assets/20819543/c600cb6f-c593-4a68-adb6-df8bbc66531c)

**Comment by erri120** [comment]
@Slayer5934 can you try one of the latest builds at https://github.com/Nexus-Mods/NexusMods.App/actions/runs/8551545718 and check the logs afterwards?

**Comment by Slayer5934** [comment]
> @Slayer5934 can you try one of the latest builds at https://github.com/Nexus-Mods/NexusMods.App/actions/runs/8551545718 and check the logs afterwards?

i will see

**Comment by Slayer5934** [comment]
[nexusmods.app.current.log](https://github.com/Nexus-Mods/NexusMods.App/files/14902403/nexusmods.app.current.log)

**Comment by erri120** [comment]
```
00:00:04.422 [INFO] (NexusMods.CrossPlatform.Process.ProcessFactory) Executing command `xdg-open steam://rungameid/413150`
00:01:04.809 [INFO] (NexusMods.DataModel.ToolManager) Ingesting loadout 393DE461B37E144BB2457E56E9F2B829 from Stardew Valley 1.6.3.24087
```

It looks like we tried to start the process and then waited for a minute for it to start. For some reason, either `steam` or the `reaper` process didn't start within a minute.

**Comment by erri120** [comment]
I've got a PR #1206 open that should improve this, or at least give us more information.

**Comment by erri120** [comment]
Can you try running https://github.com/Nexus-Mods/NexusMods.App/actions/runs/8646672224

**Comment by Slayer5934** [comment]
[nexusmods.app.main.current.log](https://github.com/Nexus-Mods/NexusMods.App/files/14973037/nexusmods.app.main.current.log)
app log ^

[SystemJournal.txt](https://github.com/Nexus-Mods/NexusMods.App/files/14973040/SystemJournal.txt)
system log for extra info ^

**Comment by erri120** [comment]
@Slayer5934 looks like debug logging wasn't enabled. I've fixed that here: https://github.com/Nexus-Mods/NexusMods.App/actions/runs/8687979869

Can you explain what actually happens on screen? What's supposed to happen is,

1) we start `xdg-open steam://rungameid/413150`,
2) Steam should open and launch the game,
3) we wait for the `steam` process to appear,
4) we wait for the `reaper` process to appear,
5) we wait until the `reaper` process exists.

**Comment by Slayer5934** [comment]
> @Slayer5934 looks like debug logging wasn't enabled. I've fixed that here: https://github.com/Nexus-Mods/NexusMods.App/actions/runs/8687979869
> 
> Can you explain what actually happens on screen?

I will download and run the new one,
When I press launch nothing happens, the only indicator it tried to do anything is it saying launching

**Comment by Slayer5934** [comment]
[nexusmods.app.main.current.log](https://github.com/Nexus-Mods/NexusMods.App/files/15003473/nexusmods.app.main.current.log)

**Comment by erri120** [comment]
```
00:00:07.244 [INFO] (NexusMods.Abstractions.Games.RunGameTool) Starting Run Stardew Valley
00:00:07.268 [INFO] (NexusMods.CrossPlatform.Process.ProcessFactory) Executing command `xdg-open steam://rungameid/413150`
00:00:07.272 [DEBUG] (NexusMods.Abstractions.Games.RunGameTool) Waiting for process `steam` to start within `0:05:00` second(s)
00:00:07.279 [DEBUG] (NexusMods.Abstractions.Games.RunGameTool) Waiting for process `reaper` to start within `0:05:00` second(s)
00:05:07.401 [WARN] (NexusMods.Abstractions.Games.RunGameTool) Process `reaper` failed to start within `0:05:00` second(s)
```

I think that `xdg-open steam://rungameid/413150` might not do anything. How did you install Steam? Are you using the Steam flatpak or something else?

**Comment by Slayer5934** [comment]
![image](https://github.com/Nexus-Mods/NexusMods.App/assets/20819543/daaa2d0e-1adf-4ffe-a8f2-93c55be77587)

**Comment by erri120** [comment]
https://github.com/Nexus-Mods/NexusMods.App/issues/1151#issuecomment-2063447963 @l0b0 can you run these commands as well?

**Comment by l0b0** [comment]
> @Slayer5934 can you run the following commands in your terminal:
> 
>     * `echo $SHELL` (expected: either `/usr/bin/bash` or `/usr/bin/zsh`)
> 
>     * `which disown` (expected: `disown: shell built-in command`)
> 
>     * `which xdg-open` (expected: `/usr/bin/xdg-open`)
> 
>     * `xdg-open --version` (expected: `xdg-open 1.2.1`)



> [#1151 (comment)](https://github.com/Nexus-Mods/NexusMods.App/issues/1151#issuecomment-2063447963) @l0b0 can you run these commands as well?

```console
❯ echo $SHELL
/run/current-system/sw/bin/bash
❯ which disown
which: no disown in (/nix/store/[omitted giant PATH])
❯ which xdg-open
/run/current-system/sw/bin/xdg-open
❯ xdg-open --version
xdg-open 1.2.1
```

**Comment by erri120** [comment]
@l0b0 and @Slayer5934 how did you install Steam? Did you use flatpak, your system package manager, or something else? Also, can you check if links in diagnostics or in the changelog are clickable? This might either be an issue with how we use `xdg-open` in general or just Steam.

I'm really perplexed by this issue because `xdg-open` appears to be working fine if you run it in a normal terminal but fails when we spawn a process with the same command.

If there isn't anything obvious that we've somehow missed, I'll probably rework how we do this entirely.

**Comment by l0b0** [comment]
> […] how did you install Steam? Did you use flatpak, your system package manager, or something else?

I basically just set [`programs.steam.enable = true;`](https://search.nixos.org/options?channel=unstable&show=programs.steam.enable&from=0&size=50&sort=relevance&type=packages&query=programs.steam.enable) in [my NixOS configuration](https://gitlab.com/engmark/root/-/blob/master/modules/gaming.nix?ref_type=heads#L44).

> Also, can you check if links in diagnostics or in the changelog are clickable? This might either be an issue with how we use `xdg-open` in general or just Steam.

[TODO]
 
> I'm really perplexed by this issue because `xdg-open` appears to be working fine if you run it in a normal terminal but fails when we spawn a process with the same command.

Doesn't `xdg-open` rely on variables such as `PATH` and `XDG_…`,  to determine either which program to run, or where to find those programs? Could it be that those are changed/unset in the context of NMA?

**Comment by Sewer56** [comment]
```
 ---> System.ComponentModel.Win32Exception (8): An error occurred trying to start process '/home/victor/.local/share/Steam/steamapps/common/Skyrim Special Edition/skse64_loader.exe' with working directory '/home/victor/.local/share/Steam/steamapps/common/Skyrim Special Edition'. Exec format error
 
 ```

I'm presuming it tried to start a Windows binary as a regular Linux one, but Wine is not present or set as the default handler, i.e. `sudo xdg-mime default`.

More importantly, I'd guess we're probably not handling running of tools (SKSE) via a WINEPREFIX, regardless of whether it's Proton or Wine.

That's a problem yet to be resolved. Skyrim is also currently unsupported and hasn't been touched since an eternity ago.

Try Stardew Valley for the time being, see how that goes.

**Comment by l0b0** [comment]
> That's a problem yet to be resolved. Skyrim is also currently unsupported and hasn't been touched since an eternity ago.
> 
> Try Stardew Valley for the time being, see how that goes.

I don't have that game, so I don't think I'll be able to contribute much more to this issue.

**Comment by erri120** [comment]
> Doesn't `xdg-open` rely on variables such as `PATH` and `XDG_…`,  to determine either which program to run, or where to find those programs? Could it be that those are changed/unset in the context of NMA?

@l0b0 We inherit the environment variables from the environment the App is running in. If you're starting the App from the shell, then it would inherit all environment variables that are present in the current shell session.

**Comment by Greg-Nexus** [comment]
[nexusmods.app.main.current.log](https://github.com/user-attachments/files/15550167/nexusmods.app.main.current.log)

I'm able to reproduce this on my Steam Deck, I think this is the correct log file.

**Comment by erri120** [comment]
This issue might be fixed with #1533.

**Comment by erri120** [comment]
Fixed with #1533, confirmed by two testers.
